stress/client/main.go

```diff
@@ -0,0 +1,322 @@
+	gauges map[string]*gauge
+}
+
+// newMetricsServer returns a new metrics server.
+func newMetricsServer() *server {
+	return &server{gauges: make(map[string]*gauge)}
+}
+
+// GetAllGauges returns all gauges.
+func (s *server) GetAllGauges(in *metricspb.EmptyMessage, stream metricspb.MetricsService_GetAllGaugesServer) error {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	for name, gauge := range s.gauges {
+		err := stream.Send(&metricspb.GaugeResponse{Name: name, Value: &metricspb.GaugeResponse_LongValue{gauge.get()}})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetGauge returns the gauge for the given name.
+func (s *server) GetGauge(ctx context.Context, in *metricspb.GaugeRequest) (*metricspb.GaugeResponse, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if g, ok := s.gauges[in.Name]; ok {
+		return &metricspb.GaugeResponse{Name: in.Name, Value: &metricspb.GaugeResponse_LongValue{g.get()}}, nil
+	}
+	return nil, grpc.Errorf(codes.InvalidArgument, "gauge with name %s not found", in.Name)
+}
+
+// createGauge creates a guage using the given name in metrics server
+func (s *server) createGauge(name string) *gauge {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if _, ok := s.gauges[name]; ok {
+		// gauge already exists.
+		panic(fmt.Sprintf("gauge %s already exists", name))
+	}
+	var g gauge
+	s.gauges[name] = &g
+	return &g
+}
+
+func startServer(server *server, port int) {
+	lis, err := net.Listen("tcp", ":"+strconv.Itoa(port))
+	if err != nil {
+		grpclog.Fatalf("failed to listen: %v", err)
+	}
+
+	s := grpc.NewServer()
+	metricspb.RegisterMetricsServiceServer(s, server)
+	s.Serve(lis)
+
+}
+
+// stressClient defines client for stress test.
+type stressClient struct {
+	testID        int
+	address       string
+	selector      *weightedRandomTestSelector
+	interopClient testpb.TestServiceClient
+	stop          <-chan bool
+}
+
+// newStressClient construct a new stressClient.
+func newStressClient(id int, addr string, conn *grpc.ClientConn, selector *weightedRandomTestSelector, stop <-chan bool) *stressClient {
+	client := testpb.NewTestServiceClient(conn)
+	return &stressClient{testID: id, address: addr, selector: selector, interopClient: client, stop: stop}
+}
+
+// mainLoop uses weightedRandomTestSelector to select test case and runs the tests.
+func (c *stressClient) mainLoop(gauge *gauge) {
+	var numCalls int64
+	timeStarted := time.Now()
+	for {
+		done := make(chan bool)
+		go func() {
+			test := c.selector.getNextTest()
+			switch test {
+			case "empty_unary":
+				interop.DoEmptyUnaryCall(c.interopClient)
+			case "large_unary":
+				interop.DoLargeUnaryCall(c.interopClient)
+			case "client_streaming":
+				interop.DoClientStreaming(c.interopClient)
+			case "server_streaming":
+				interop.DoServerStreaming(c.interopClient)
+			case "empty_stream":
+				interop.DoEmptyStream(c.interopClient)
+			default:
+				grpclog.Fatalf("Unsupported test case: %d", test)
+			}
+			done <- true
```

Comment:
This could be stuck forever if c.stop is read in line 251. Make done with buffer 1.